### PR TITLE
Refactor `scatter.cuh` and `gather.cuh`

### PR DESCRIPTION
**Pull request by ttnghia**
This PR refactors `scatter.cuh` and `gather.cuh`, totally moving all the implementations of the templated functions `detail::scatter` and `detail::gather` into separates `*.cu` files with explicit instantiations.

Typically, this is not a trivial task since the APIs `scatter` and `gather` take templated iterator parameters (iterators for scatter/gather maps), and the types of those iterators can be very complex templated `thrust::transform_iterator`. In this PR, in order to explicitly instantiate the `scatter` and `gather` APIs, I have to limit the iterators types to be only pointers and `thrust::constant_iterator`. For `thrust::transform_iterator`, I create a temporary `device_uvector` and fill it with the transformed iterator values then pass its `begin()/end()` iterators to the `scatter` and `gather` APIs. Since most cudf APIs are passing scatter/gather maps from a `device_uvector` iterators, I don't think this would incur much overhead. The files affected by this are here: https://github.com/rapidsai/cudf/pull/8464/commits/37366cfcd2e3cac8b89ac751137d8898fd28758c.

After doing this (rough) refactoring, I gained about 5% speedup in compile time (on a 18 cores machine, 64 compile threads):
Before:
```
cudf: 6:48s
Others: 7:19s
All: 13:13s
```
After:
```
cudf: 6:31s
Others: 7:19s
All: 12:55s
```

It seems that 5% is not much worth do this, but if we merge this (after more cleanup) we will have a very nice scatter and gather APIs implementations: the header files `*.cuh` contain only declarations, and the implementations will be freed from including into many many others `*cu` file. Thus, I need more opinions from you guys about this before spending more time working on it.

-----------
### PS: This is not a complete work, just a rough draft and still needs much cleanup.

### CODE REVIEW

**Comment by jrhemstad** [comment]
I am not a fan of requiring materializing any iterator based gather or scatter map into a `device_uvector`.

**Comment by ttnghia** [comment]
> I am not a fan of requiring materializing any iterator based gather or scatter map into a `device_uvector`.

Me too. However, most (maybe 90% or so) cudf APIs already did it.

**Comment by harrism** [comment]
THis should probably be a draft PR.

File: cpp/src/copying/copy.cu [hunk 1]
```diff
@@ -200,10 +200,13 @@ std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
     auto const gather_lhs = make_counting_transform_iterator(
       size_type{0}, lhs_gather_map_functor<Filter>{is_left, null_map_entry});
 
+    rmm::device_uvector<size_type> gather_map(size, stream);
+    thrust::copy(rmm::exec_policy(stream), gather_lhs, gather_lhs + size, gather_map.begin());
```
Comment (harrism) [inline cpp/src/copying/copy.cu:204]: Yuck.

**Comment by harrism** [comment]
I don't like the idea of forcing materialization. That kills the benefit of an iterator interface.

`cudf::reverse` is an example of a case where we don't need to materialize the gather map. https://github.com/rapidsai/cudf/pull/8410

**Comment by ttnghia** [comment]
> THis should probably be a draft PR.

I just tried to convert it to a draft but was told that "subscribed people will be unsubscribed" :smile:, so just keep a normal (WIP) PR to let other participants know the activities here.

**Comment by ttnghia** [comment]
> I don't like the idea of forcing materialization. That kills the benefit of an iterator interface.
> 
> `cudf::reverse` is an example of a case where we don't need to materialize the gather map. #8410

Yeap, this is a trade-off.

Based on the opinions, it seems that we should not continue doing this...

**Pull request closed without merging**